neps/nep-0508.md

```diff
@@ -0,0 +1,264 @@
+
+### New shard layout
+
+The first release of the resharding v2 will contain a new shard layout where one of the existing shards will be split into two smaller shards. Furthermore additional reshardings can be scheduled in subsequent releases without additional NEPs unless the need for it arises. A new shard layout can be determined and will be scheduled and executed with the next protocol upgrade. Resharding will typically happen by splitting one of the existing shards into two smaller shards. The new shard layout will be created by adding a new boundary account that will be determined by analysing the storage and gas usage metrics within the shard and selecting a point that will divide the shard roughly in half in accordance to the mentioned metrics. Other metrics can also be used based on requirements.
+
+### Removal of Fixed shards
+
+Fixed shards was a feature of the protocol that allowed for assigning specific accounts and all of their recursive sub accounts to a predetermined shard. This feature was only used for testing and was never used in production. Fixed shards feature unfortunately breaks the contiguity of shards and is not compatible with the new resharding flow. A sub account of a fixed shard account can fall in the middle of account range that belongs to a different shard. This property of fixed shards made it particularly hard to reason about and implement efficient resharding. 
+
+For example in a shard layout with boundary accounts [`b`, `d`] the account space is cleanly divided into three shards, each spanning a contiguous range and account ids:
+
+* 0 -  `:b`
+* 1 - `b:d`
+* 2 - `d:`
+
+Now if we add a fixed shard `f` to the same shard layout, then any we'll have 4 shards but neither is contiguous. Accounts such as `aaa.f`, `ccc.f`, `eee.f` that would otherwise belong to shards 0, 1 and 2 respectively are now all assigned to the fixed shard and create holes in the shard account ranges.  
+
+It's also worth noting that there is no benefit to having accounts colocated in the same shard. Any transaction or receipt is treated the same way regardless of crossing shard boundary. 
+
+This was implemented ahead of this NEP and the fixed shards feature was **removed**. 
+
+### Garbage collection
+
+In epoch T+2 once resharding is completed, we can delete the trie state and the flat state related to the parent shard. In practice, this is handled as part of the garbage collection code. While garbage collecting the last block of epoch T+1, we go ahead and clear all the data associated with the parent shard from the trie cache, flat storage, and RocksDB state associated with trie state and flat storage.
```

Comment:
Does this mean that the storage increase would last for 5 epochs (or however long the garbage collection period is configured to be)

Reply:
Yes, that's correct, we maintain the state of the parent shard for the period of GC. Note that right after resharding, we keep deleting state from the `State` (Trie data) column of the parent shard with GC normally while new state is being added to the `State` column of the child shards upto the GC period. This means the storage increase compared to a non-resharding case would amortized be the size of the current state as of the time of resharding and not all the historic data.